packages/asset-swapper/src/network/chain.ts

```diff
@@ -0,0 +1,314 @@
+import { BigNumber, hexUtils } from '@0x/utils';
+import * as crypto from 'crypto';
+
+import { CallDispatcherContract } from '../wrappers';
+import { artifacts } from '../artifacts';
+
+import { DUMMY_PROVIDER, NULL_BYTES, ZERO_AMOUNT } from './constants';
+import { Address, Bytes } from './types';
+
+export interface ChainEthCallOverrides {
+    [address: string]: {
+        code?: Bytes;
+    },
+};
+
+export interface ChainEthCallOpts {
+    data: Bytes;
+    to: Address;
+    value?: BigNumber;
+    overrides?: ChainEthCallOverrides;
+    immediate?: boolean;
+    maxCacheAgeMs?: number;
+    gas?: number;
+    gasPrice?: BigNumber;
+}
+
+interface QueuedEthCall {
+    opts: ChainEthCallOpts,
+    accept: (v: Bytes) => void;
+    reject: (v: string | Bytes) => void;
+}
+
+export interface CreateChainOpts {
+    provider: SupportedProvider;
+    tickFrequency?: number;
+    maxCacheAgeMs?: number;
+}
+
+const DEFAULT_CALL_GAS_LIMIT = 10e6;
+const DEFAULT_CALLER_ADDRESS = hexUtils.random(20);
+const DISPATCHER_CONTRACT_ADDRESS = hexUtils.random(20);
+const DISPATCHER_CONTRACT = new CallDispatcherContract(
+    DISPATCHER_CONTRACT_ADDRESS,
+    DUMMY_PROVIDER,
+);
+const DISPATCHER_CONTRACT_BYTECODE = artifacts.CallDispatcher.compilerOutput.evm.deployedBytecode.object;
+
+export interface DispatchedCallResult {
+    success: boolean;
+    resultData: Bytes;
+}
+
+interface CachedDispatchedCallResult {
+    result: DispatchedCallResult;
+    cacheTimeMs: number;
+}
+
+export interface Chain {
+    chainId: ChainId;
+    provider: SupportedProvider;
+    ethCall(opts: ChainEthCallOpts): Promise<Bytes>;
+}
+
+export class LiveChain implements Chain {
+    private readonly _w3: Web3Wrapper;
+    private readonly _chainId: ChainId;
+    private readonly _cachedCallResults: { [id: string]: CachedDispatchedCallResult } = {};
+    private _maxCacheAgeMs: number = 0;
+    private _tickTimer: NodeJS.Timeout | null = null;
+    private _queue: QueuedEthCall[] = [];
+
+    public static async createAsync(opts: CreateChainOpts): Promise<Chain> {
+        const _opts = { ...opts, tickFrequency: 100, maxCacheAgeMs: 30e3 };
+        const w3 = new Web3Wrapper(opts.provider);
+        const chainId = await w3.getChainIdAsync() as ChainId;
+        const inst = new LiveChain(chainId, w3, _opts.maxCacheAgeMs);
+        await inst._start(_opts.tickFrequency);
+        return inst;
+    }
+
+    protected constructor(chainId: number, w3: Web3Wrapper, maxCacheAgeMs: number) {
+        this._chainId = chainId;
+        this._w3 = w3;
+        this._maxCacheAgeMs = maxCacheAgeMs;
+    }
+
+    protected async _start(tickFrequency: number): Promise<void> {
+        await this._tick(tickFrequency, true);
+    }
+
+    public get provider(): SupportedProvider {
+        return this._w3.getProvider();
+    }
+
+    public get chainId(): ChainId {
+        return this._chainId;
+    }
+
+    public async ethCall(opts: ChainEthCallOpts): Promise<Bytes> {
```

Comment:
People call this function to queue up a call that'll get batched up and executed after every tick